azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/data/CallHistoryRepository.kt

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.calling.data
+
+import android.content.ContentValues
+import android.content.Context
+import com.azure.android.communication.ui.calling.data.model.CallHistoryRecordData
+import org.threeten.bp.LocalDateTime
+import org.threeten.bp.format.DateTimeFormatter
+
+internal interface CallHistoryRepository {
+    fun insert(callId: String, callDateTime: LocalDateTime)
+    fun getAll(): List<CallHistoryRecordData>
+    fun remove(id: Int)
+}
+
+internal class CallHistoryRepositoryImpl(
+    private val context: Context
+) : CallHistoryRepository {
+
+    private val datePattern: String = "yyyy-MM-dd HH:mm:ss"
+
+    override fun insert(callId: String, callDateTime: LocalDateTime) {
+        val db = DbHelper(context).writableDatabase
```

Comment:
The way you're using DbHelper isn't optimal here. Since it extends `SQLiteOpenHelper`, it will handle things like caching of an opened db for you.

You're creating this instance, and throwing it away right after. Which means you're throwing away your "opened database cache".

So, every time you're accessing `CallHistoryRepositoryImpl`, you're going through the entire "open database on disk, run the CREATE statements if necessary, check its version pragma, determine if we need to upgrade, etc" logic. A bunch of slow and unnecessary work.

Instead, consider storing `DbHelper(context)` as a variable on this class. This way you're only incurring the costs of opening/setting up the database once.

Also, you can then control when you incur the db open/setup costs - when this class is initialized (e.g., touch `writableDatabase` in an `init {}` block), or on a first interaction, etc.

You'll also need to determine when you want to close this database, as well - if at all. Probably when `CallCompositeActivity` is being torn down?

Reply:
As per our discussion, decided to keep current logic as it is a small tradeoff in comparison of risk to keep DB connection open, as we do not have a reliable close event.